### PR TITLE
feat: Delete rows and columns infrastructure

### PR DESCRIPTION
**Pull request by tihuan**
### Reviewers

**Functional:** 
@seve 

**Readability:** 
@MDunitz 

---

## Changes
- add
1. `src/views/WheresMyGene/common/store` for `useReducer` to centralize heat map state, since many values change together in various events
2. The reducer/actions pattern follows closely to how [Redux Toolkit](https://redux-toolkit.js.org/usage/usage-with-typescript#createslice) structures the code to reduce boilerplate. I didn't dynamically generate actions like RTK does because the TS overhead involved to type things properly would take a lot more time for me to figure out 😆 And I don't think the app is global/heavy enough to justify adding RTK, so hopefully the current approach is a good balance
3. Add `useDeleteGenesAndCellTypes()` hook to encapsulate business logic, which internally uses the reducer and dispatch relevant actions as needed
4. Add `ehcarts` click events for row and column labels. Also update its style to highlight labels when selected
- remove
- modify
1. Move `selectedGenes` to use the store state

## QA steps (optional)

1. Load some genes
2. You can now select/deselect rows and columns! NOTE: I looked into highlighting rows and columns on selection as shown in [Figma](https://www.figma.com/file/htICFl7zihfs1z9eIYVwzw/Gene-Search?node-id=2442%3A430111), but couldn't find a quick way to do it, so decided to just highlight the labels for now
3. I haven't added delete key press detection yet 😆 
4. If you delete a gene, it will deselect the labels

demo

### CODE REVIEW

File: frontend/src/views/WheresMyGene/components/HeatMap/index.tsx [hunk 1]
```diff
@@ -19,6 +21,7 @@ import {
   createXAxisOptions,
   createYAxisOptions,
   dataToChartFormat,
+  deserializeCellTypeName,
```
Comment (tihuan) [inline frontend/src/views/WheresMyGene/components/HeatMap/index.tsx:24]: helper function to deserialize event value for better developer experience

File: frontend/src/views/WheresMyGene/components/HeatMap/utils.ts [hunk 2]
```diff
@@ -154,6 +162,17 @@ export function createXAxisOptions({
     xAxis: [
       {
         axisLabel: {
+          formatter(value) {
```
Comment (tihuan) [inline frontend/src/views/WheresMyGene/components/HeatMap/utils.ts:165]: add highlight style to gene labels when selected

File: frontend/src/views/WheresMyGene/components/HeatMap/utils.ts [hunk 3]
```diff
@@ -225,7 +250,23 @@ export function createYAxisOptions({
     ],
     yAxis: [
       {
-        axisLabel: { align: "right", rotate: 20 },
+        axisLabel: {
```
Comment (tihuan) [inline frontend/src/views/WheresMyGene/components/HeatMap/utils.ts:253]: add highlight style to cell type labels when selected

File: frontend/src/views/WheresMyGene/components/HeatMap/utils.ts [hunk 4]
```diff
@@ -326,10 +368,43 @@ export function getHeatmapHeight(cellTypes: CellTypeSummary[]): number {
...
 
+/**
+ * Value format: `${id}~${tissue}~${name}`
+ * We need to encode cell type metadata here, so we can use it in onClick event
+ */
 export function getCellTypeNames(cellTypes: CellTypeSummary[]): string[] {
-  return cellTypes.map((cellType) => cellType.name);
+  return cellTypes.map(({ id, name }) => {
+    // TODO(thuang): Add multiple tissues support.
+    return `${id}~lung~${name}`;
```
Comment (tihuan) [inline frontend/src/views/WheresMyGene/components/HeatMap/utils.ts:378]: Hard coding `lung` right now, but in the next PR I'll add more multiple tissue support!

File: frontend/src/views/WheresMyGene/index.tsx [hunk 5]
```diff
@@ -25,31 +32,34 @@ import { Wrapper } from "./style";
...
+  const { selectedGenes } = state;
 
   /**
    * This holds ALL the geneData we have loaded from the API, including previously
    * and currently selected genes.
    * We use `selectedGeneData` to subset the data to only the genes that are
    * currently selected.
    */
-  const [geneData, setGeneData] =
+  const [geneExpressionSummaries, setGeneExpressionSummaries] =
```
Comment (tihuan) [inline frontend/src/views/WheresMyGene/index.tsx:45]: just name change to better convey the meaning

File: frontend/src/views/WheresMyGene/index.tsx [hunk 6]
```diff
@@ -25,31 +32,34 @@ import { Wrapper } from "./style";
...
     useState<GeneExpressionSummary[]>(EMPTY_ARRAY);
+
   const [cellTypes, setCellTypes] = useState<CellTypeSummary[]>(EMPTY_ARRAY);
 
   /**
    * This is the formatted data that we use to render the heatmap.
    */
-  const [data, setData] = useState<CellTypeSummary[]>(EMPTY_ARRAY);
+  const [cellTypeSummaries, setCellTypeSummaries] =
```
Comment (tihuan) [inline frontend/src/views/WheresMyGene/index.tsx:53]: name change only

**Review by tihuan — COMMENTED** [review]
Hopefully the comments are helpful! Thank you!

**Review by MDunitz — APPROVED** [review]
lgtm

**Review by seve — APPROVED** [review]
LGTM! Looks awesome

**Comment by tihuan** [comment]
Fantastic!! Thanks so much, both 🤩 🙏 !!

**Pull request merged**